pool-manager.js

```diff
@@ -0,0 +1,132 @@
+var _ = require('underscore');
+
+function PoolManager (options) {
+  options = options || {};
+
+  this.endpoints = [];
+  this._endpointOffset = 0;
+
+  this.isInPool = options.isInPool || _.constant(true);
+  this.onEndpointReturned = options.onEndpointReturned || _.noop;
+  this.onEndpointRegistered = options.onEndpointRegistered || _.noop;
+}
+
+PoolManager.prototype = {
+  getNextEndpoint: function () {
+    var i, l, offset, endpoint;
+
+    for (i = 0, l = this.endpoints.length; i < l; ++i) {
+      offset = (this._endpointOffset + i) % l;
+      endpoint = this.endpoints[offset];
+
+      if (this.isInPool(endpoint)) {
+        this._endpointOffset = offset + 1;
+        return endpoint;
+      }
+    }
+  },
+  updateEndpoints: function (endpoints) {
+    newEndpoints = endpoints.map(function (info) {
+      return new Endpoint(info);
+    });
+
+    for (i = this.endpoints.length; i--;) {
+      matchingEndpoint = _.findWhere(newEndpoints, { url: this.endpoints[i].url });
+
+      if (matchingEndpoint) { // found a match, remove it from `newEndpoints`, since it's not new
+        newEndpoints = _.without(newEndpoints, matchingEndpoint);
+      } else { // didn't find a match in endpoints, so kill that endpoint
+        this.endpoints.splice(i, 1);
+      }
+    }
+
+    newEndpoints.forEach(function (endpoint) {
+      endpoint.callback = this.onEndpointReturned.bind(this, endpoint);
+      this.onEndpointRegistered(endpoint);
+    }, this);
+    // push all the actually-new endpoints in
+    this.endpoints.push.apply(this.endpoints, newEndpoints);
+  }
+};
+
+function Endpoint(info) {
+  this.name = info.name;
+  this.port = info.port;
+  this.url = info.name + ':' + info.port;
+}
+
+module.exports = {
+  defaultPoolManager: function () {
+    return new PoolManager();
+  },
+  ejectOnErrorPoolManager: function (options) {
+    // endpoint states
+    var CLOSED            = 0,  // closed circuit: endpoint is good to use
+        HALF_OPEN_READY   = 1,  // endpoint is in recovery state: offer it for use once
+        HALF_OPEN_PENDING = 2,  // endpoint recovery is in process
+        OPEN              = 3;  // open circuit: endpoint is no good
+
+    if (!options || !(options.failureWindow && options.maxFailures && options.resetTimeout)) {
+      throw new Error('Must supply all arguments to ejectOnErrorPoolManager');
+    }
+
+    var failureWindow = options.failureWindow;
+    var maxFailures = options.maxFailures;
+    var resetTimeout = options.resetTimeout;
+
+    function disableEndpoint(endpoint) {
+      endpoint.state = OPEN;
```

Comment:
The point of not doing it that way was because `state` only matters within the scope of this `ejectOnErrorPoolManager` - a regular pool manager doesn't care about this.